cdci_data_analysis/analysis/ontology.py

```diff
@@ -0,0 +1,364 @@
+                if len(possible_datatypes) == 1:
+                    limits_datatype = list(possible_datatypes)[0]
+            
+            lim_r = []
+            if len(ll) != 0:
+                lim_r.append(rdf.BNode())
+                graph.add((lim_r[-1], 
+                           XSD.minInclusive, 
+                           rdf.Literal(xsd_type_to_python_type(limits_datatype)(ll[0].value), 
+                                       datatype=limits_datatype)))
+            if len(ul) != 0:
+                lim_r.append(rdf.BNode())
+                graph.add((lim_r[-1], 
+                           XSD.maxInclusive, 
+                           rdf.Literal(xsd_type_to_python_type(limits_datatype)(ul[0].value), 
+                                       datatype=limits_datatype)))
+            c = Collection(graph, None, lim_r)
+                        
+            dtype = rdf.BNode()
+            graph.add((dtype, a, RDFS.Datatype))
+            graph.add((dtype, OWL.onDatatype, limits_datatype))
+            graph.add((dtype, OWL.withRestrictions, c.uri))
+            
+            bn = rdf.BNode()
+            graph.add((bn, a, OWL.Restriction))
+            graph.add((bn, OWL.onProperty, ODA.value))
+            graph.add((bn, OWL.allValuesFrom, dtype))
+            
+            graph.add((classuri, RDFS.subClassOf, bn))
+    
+    def _get_datatype_restriction(self, param_uri):
+        param_uri = f"<{param_uri}>" if param_uri.startswith("http") else param_uri
+        query = """
+            SELECT ?dt WHERE {
+                {
+                    %s rdfs:subClassOf+ [
+                        a owl:Restriction ;
+                        owl:onProperty oda:value ;
+                        owl:allValuesFrom ?dt
+                    ]   
+                    FILTER(isUri(?dt) && STRSTARTS(STR(?dt), STR(xsd:)))
+                }
+                UNION
+                {
+                    BIND(%s as ?dt)
+                    FILTER(STRSTARTS(STR(%s), STR(xsd:)))
+                }
+                UNION
+                {
+                    %s rdfs:subClassOf+ ?dt .
+                    FILTER(isUri(?dt) && STRSTARTS(STR(?dt), STR(xsd:)))
+                }
+            }
+        """ % (param_uri, param_uri, param_uri, param_uri)
+        qres = list(self.g.query(query))
+        if len(qres) == 0: return None
+        if len(set(r[0] for r in qres)) > 1:
+            raise RuntimeError("Ambiguous datatype of %s", param_uri) 
+        return qres[0][0]
+    
+    def parse_extra_ttl(self, extra_ttl, parse_oda_annotations = True):
+        if parse_oda_annotations:
+            tmpg = rdf.Graph()
+            tmpg.parse(data = extra_ttl)       
+            try:
+                self.parse_oda_annotations(tmpg)
+            except RuntimeError as e:
+                raise RequestNotUnderstood(e.message)
+            extra_ttl = tmpg.serialize(format='turtle')
```

Comment:
In the plugin I have till now used `n3` which is basically a superset of ttl, perhaps we want to have them aligned, even though nothing should change

Reply:
here, the serialisation is used only internally, to feed it to `self.g.parse()` (in the next line), so any format would generally work, I use turtle as it is the default expected format for `parse` 
Also, this method is used to parse `extra_ttl` from nb2workflow, which is turtle.

(But I will extend this method a bit, so that any valid triple format can be used)